presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java

```diff
@@ -147,6 +147,87 @@ public void testDistributesSpillOverPathsBadDisk()
         assertEquals(listFiles(spillPath2.toPath()).size(), 0);
     }
 
+    @Test
+    public void testCacheInvalidatedOnBadDisk()
+            throws Exception
+    {
+        List<Type> types = ImmutableList.of(BIGINT);
+        List<Path> spillPaths = ImmutableList.of(spillPath1.toPath(), spillPath2.toPath());
+        FileSingleStreamSpillerFactory spillerFactory = new FileSingleStreamSpillerFactory(
+                executor, // executor won't be closed, because we don't call destroy() on the spiller factory
+                blockEncodingSerde,
+                new SpillerStats(),
+                spillPaths,
+                1.0,
+                false,
+                false);
+
+        assertEquals(listFiles(spillPath1.toPath()).size(), 0);
+        assertEquals(listFiles(spillPath2.toPath()).size(), 0);
+
+        Page page = buildPage();
+        List<SingleStreamSpiller> spillers = new ArrayList<>();
+        for (int i = 0; i < 2; ++i) {
+            SingleStreamSpiller singleStreamSpiller = spillerFactory.create(types, bytes -> {}, newSimpleAggregatedMemoryContext().newLocalMemoryContext("test"));
+
+            if (i == 1) {
+                // Set second spiller path to read-only after initialization to emulate a disk failing during runtime
+                setPosixFilePermissions(spillPath2.toPath(), ImmutableSet.of(PosixFilePermission.OWNER_READ));
+            }
+
+            try {
+                getUnchecked(singleStreamSpiller.spill(page));
+            }
+            catch (Exception ignored) {
+                // Do nothing because we expect this to fail due to file permissions
```

Comment:
Please use `org.assertj.core.api.AssertionsForClassTypes#assertThatThrownBy(org.assertj.core.api.ThrowableAssert.ThrowingCallable)` 
 to verify expected exceptions.